File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/渐进圆环效果/WGCircleAniView.m
//
//  WGCircleAniView.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGCircleAniView.h"

@interface WGCircleAniView()<CAAnimationDelegate>

@property (nonatomic, strong) CAShapeLayer *shapeLayer;

@end

@implementation WGCircleAniView

- (instancetype)initWithFrame:(CGRect)frame{
    
    self = [super initWithFrame:frame];
    if (self) {
        
        [self buildBaseView];
    }
    return self;
}

- (void)buildBaseView{
    
    self.shapeLayer = [[CAShapeLayer alloc] init];
    
    UIBezierPath *circlePath = [UIBezierPath bezierPath];
    [circlePath addArcWithCenter:self.center radius:self.height/2-2.5 startAngle:0 endAngle:M_PI*2 clockwise:YES];
    
    _shapeLayer.path = circlePath.CGPath;
    _shapeLayer.lineWidth = 5;
    _shapeLayer.lineCap = kCALineCapRound;
    _shapeLayer.fillColor = [UIColor clearColor].CGColor;
    [self.layer addSublayer:_shapeLayer];
}

- (void)startCircleAniColor:(UIColor *)color duation:(CGFloat)duation;{
    
    self.hidden = NO;
    self.shapeLayer.strokeColor = color.CGColor;
    
    CABasicAnimation *ani = [[CABasicAnimation alloc] init];
    ani.keyPath = @"strokeEnd";
    ani.duration = duation;
    ani.fromValue = @(0);
    ani.toValue = @(1);
    ani.delegate = self;
    ani.fillMode = kCAFillModeForwards;
    [ani setRemovedOnCompletion:NO];
    [self.shapeLayer addAnimation:ani forKey:nil];
}

-(void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag{
    
    if (flag) {
        //进度动画完成,回调
        if (self.circleAniViewDelegate && [self.circleAniViewDelegate respondsToSelector:@selector(aniDidStop:finished:)]) {
            
            self.hidden = YES;
            [self.circleAniViewDelegate aniDidStop:anim finished:flag];
        }
    }
}

@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/渐进圆环效果/WGCircleAniView.swift
//
//  WGCircleAniView.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/4/24.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

protocol WGCircleAniViewDelegate {
    func circleAniStop()
}

class WGCircleAniView: UIView {

    var lineWidth: CGFloat = 4
    var strokeColor = UIColor.black
    var circle = CAShapeLayer()
    var wgCircleAniViewDelegate: WGCircleAniViewDelegate?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        let startAngle = CGFloat(Double.pi)*(90.01/180.0)
        let endAngle = CGFloat(Double.pi)*(90.0/180.0)
        let circlePath = UIBezierPath(arcCenter: CGPoint(x: self.frame.size.width/2, y: self.frame.size.height/2), radius: self.frame.size.height/2-2, startAngle: startAngle, endAngle: endAngle, clockwise: true)
        circle.path = circlePath.cgPath
        circle.lineCap = .round
        circle.fillColor = UIColor.clear.cgColor
        circle.lineWidth = lineWidth
        self.layer.addSublayer(circle)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func startCircleAnimation(duation: CFTimeInterval) {
        
        circle.strokeColor = strokeColor.cgColor
        let progressAni = CABasicAnimation()
        progressAni.keyPath = "strokeEnd"
        progressAni.delegate = self
        progressAni.fromValue = 0
        progressAni.toValue = 1
        progressAni.duration = duation
        progressAni.fillMode = .forwards
        progressAni.isRemovedOnCompletion = false
        progressAni.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        progressAni.setValue("strokeEndAnimationcircle", forKey: "groupborderkeycircle")
        circle.add(progressAni, forKey: "strokeEndAnimationcircle")
    }
}

extension WGCircleAniView: CAAnimationDelegate{
    
    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        
        wgCircleAniViewDelegate?.circleAniStop()
    }
}
